File: backend/src/models/Student.js
const pool = require('../config/database');

const Student = {
  async findAll({ class_id, q } = {}) {
    let query = `
      SELECT s.*, c.label AS class_label
      FROM students s
      LEFT JOIN classes c ON s.class_id = c.id
      WHERE 1=1
    `;
    const params = [];

    if (class_id) {
      params.push(class_id);
      query += ` AND s.class_id = $${params.length}`;
    }
    if (q) {
      params.push(`%${q}%`);
      query += ` AND (s.first_name ILIKE $${params.length} OR s.last_name ILIKE $${params.length})`;
    }
    query += ' ORDER BY s.last_name, s.first_name';

    const { rows } = await pool.query(query, params);
    return rows;
  },

  async findById(id) {
    const { rows } = await pool.query(
      `SELECT s.*, c.label AS class_label
       FROM students s
       LEFT JOIN classes c ON s.class_id = c.id
       WHERE s.id = $1`,
      [id]
    );
    return rows[0] || null;
  },

  async create({ first_name, last_name, email, class_id, photo_url }) {
    const { rows } = await pool.query(
      `INSERT INTO students (first_name, last_name, email, class_id, photo_url)
       VALUES ($1, $2, $3, $4, $5) RETURNING *`,
      [first_name, last_name, email || null, class_id || null, photo_url || null]
    );
    return rows[0];
  },

  async update(id, fields) {
    const allowed = ['first_name', 'last_name', 'email', 'class_id', 'photo_url'];
    const updates = [];
    const values = [];

    for (const key of allowed) {
      if (fields[key] !== undefined) {
        values.push(fields[key]);
        updates.push(`${key} = $${values.length}`);
      }
    }

    if (updates.length === 0) return null;

    values.push(id);
    const { rows } = await pool.query(
      `UPDATE students SET ${updates.join(', ')} WHERE id = $${values.length} RETURNING *`,
      values
    );
    return rows[0] || null;
  },

  async updatePhoto(id, photo_url) {
    const { rows } = await pool.query(
      'UPDATE students SET photo_url = $1 WHERE id = $2 RETURNING *',
      [photo_url, id]
    );
    return rows[0] || null;
  },

  async delete(id) {
    const { rowCount } = await pool.query('DELETE FROM students WHERE id = $1', [id]);
    return rowCount > 0;
  },

  async bulkCreate(students) {
    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      const created = [];
      for (const s of students) {
        const { rows } = await client.query(
          `INSERT INTO students (first_name, last_name, email, class_id, photo_url)
           VALUES ($1, $2, $3, $4, $5) RETURNING *`,
          [s.first_name, s.last_name, s.email || null, s.class_id || null, s.photo_url || null]
        );
        created.push(rows[0]);
      }
      await client.query('COMMIT');
      return created;
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  },
};

module.exports = Student;

File: backend/src/models/student.py
from ..config.database import query, query_one, get_conn
from psycopg2.extras import RealDictCursor


class StudentModel:
    @staticmethod
    def find_all(class_id=None, q=None) -> list[dict]:
        sql = """
            SELECT s.*, c.label AS class_label
            FROM students s
            LEFT JOIN classes c ON s.class_id = c.id
            WHERE 1=1
        """
        params: list = []
        if class_id:
            sql += " AND s.class_id = %s"
            params.append(class_id)
        if q:
            sql += " AND (s.first_name ILIKE %s OR s.last_name ILIKE %s)"
            params.extend([f"%{q}%", f"%{q}%"])
        sql += " ORDER BY s.last_name, s.first_name"
        return query(sql, params or None)

    @staticmethod
    def find_by_id(id: int) -> dict | None:
        return query_one(
            """
            SELECT s.*, c.label AS class_label
            FROM students s
            LEFT JOIN classes c ON s.class_id = c.id
            WHERE s.id = %s
            """,
            (id,),
        )

    @staticmethod
    def create(
        first_name: str,
        last_name: str,
        email=None,
        class_id=None,
        photo_url=None,
    ) -> dict:
        return query_one(
            """
            INSERT INTO students (first_name, last_name, email, class_id, photo_url)
            VALUES (%s, %s, %s, %s, %s) RETURNING *
            """,
            (first_name, last_name, email, class_id, photo_url),
        )

    @staticmethod
    def update(id: int, fields: dict) -> dict | None:
        allowed = ["first_name", "last_name", "email", "class_id", "photo_url"]
        updates, values = [], []
        for key in allowed:
            if key in fields:
                updates.append(f"{key} = %s")
                values.append(fields[key])
        if not updates:
            return None
        values.append(id)
        return query_one(
            f"UPDATE students SET {', '.join(updates)} WHERE id = %s RETURNING *",
            values,
        )

    @staticmethod
    def update_photo(id: int, photo_url: str) -> dict | None:
        return query_one(
            "UPDATE students SET photo_url = %s WHERE id = %s RETURNING *",
            (photo_url, id),
        )

    @staticmethod
    def delete(id: int) -> bool:
        rows = query("DELETE FROM students WHERE id = %s RETURNING id", (id,))
        return len(rows) > 0

    @staticmethod
    def bulk_create(students: list[dict]) -> list[dict]:
        with get_conn() as conn:
            with conn.cursor(cursor_factory=RealDictCursor) as cur:
                created = []
                for s in students:
                    cur.execute(
                        """
                        INSERT INTO students (first_name, last_name, email, class_id, photo_url)
                        VALUES (%s, %s, %s, %s, %s) RETURNING *
                        """,
                        (
                            s["first_name"],
                            s["last_name"],
                            s.get("email"),
                            s.get("class_id"),
                            s.get("photo_url"),
                        ),
                    )
                    created.append(dict(cur.fetchone()))
                return created
